kayak_core/src/widget.rs

```diff
@@ -1,26 +1,113 @@
+use std::any::Any;
 use as_any::AsAny;
 
 use crate::{context::KayakContext, context_ref::KayakContextRef, styles::Style, Event, Index, OnEvent, Children};
 
-pub trait Widget: std::fmt::Debug + AsAny + Send + Sync {
-    /// Returns whether this widget can be focused or not (or unspecified if `None`)
-    fn focusable(&self) -> Option<bool>;
+/// An internal trait that has a blanket implementation over all implementors of [Widget]
+///
+/// This ensures that [BaseWidget] can never be implemented manually outside of this crate, even
+/// if it is exported out (as long as this one isn't).
+pub trait SealedWidget {}
+
+/// The base widget trait, used internally
+///
+/// You should _never_ implement BaseWidget manually. It is automatically implemented on
+/// all implementors of [Widget].
+pub trait BaseWidget: SealedWidget + std::fmt::Debug + Send + Sync {
+    fn constructor<P: WidgetProps>(props: P) -> Self where Self: Sized;
     fn get_id(&self) -> Index;
     fn set_id(&mut self, id: Index);
-    fn get_styles(&self) -> Option<Style>;
-    fn get_name(&self) -> String;
-    fn on_event(&mut self, context: &mut KayakContext, event: &mut Event);
+    fn get_props(&self) -> &dyn WidgetProps;
+    fn get_props_mut(&mut self) -> &mut dyn WidgetProps;
     fn render(&mut self, context: &mut KayakContextRef);
+    fn get_name(&self) -> &'static str;
+    fn on_event(&mut self, context: &mut KayakContext, event: &mut Event);
 }
 
-impl as_any::Downcast for dyn Widget {}
-impl as_any::Downcast for dyn Widget + Send {}
-impl as_any::Downcast for dyn Widget + Sync {}
-impl as_any::Downcast for dyn Widget + Send + Sync {}
+pub trait Widget: std::fmt::Debug + Clone + Default + PartialEq + AsAny + Send + Sync {
```

Comment:
Wait how can we do `+ Clone` here? Pretty sure that wasn't allowed. 🤔

Reply:
It isn't allowed if we want `Widget` to be object-safe. Using `BaseWidget` allows us to side-step that requirement by turning `Self::Props` into `dyn WidgetProps`.

In theory, this allows implementors of widget to work with their concrete types, but allows our internals to just work with the traits.